### PR TITLE
VTA Chisel Wide memory interface.

### PR DESCRIPTION
**Pull request by aasorokiin**
VTA modification for parametrizable AXI data size.

Performance change:
Numbers are based on test_benchmark_topi_conv2d.py limited to resnet-18.C2
Baseline 64bit data tsim run: 192M cycles. New 64bit data transfer: 97M cycles.
128bit data: 58M cycles. 256bit data: 39M cycles. 512bit data: 29M cycles.
Code changes:
* AXI 64/128/256/512 data bits support by AXIParams->dataBits
* TensorLoad is modified to replace all VME load operations. Multiple simultaneous requests can be generated. Load is pipelined and separated from request generation.  A "wide" implementation of load/store is used when AXI interface data width is larger than number of bits of a tensor.
* TensorStore -> TensorStoreNarrowVME TensorStoreWideVME. The narrow one is the original one
* TensorLoad -> TensorLoadSimple (original) TensorLoadWideVME TensorLoadNarrowVME
* LoadUop -> LoadUopSimple is the original one. The new one is based on TensorLoad
* Fetch -> FetchVME64 FetchWideVME. Reuse communication part from TensorLoad. Implemented as a 64bit tensor  with double tensor read to allow 64bit address alignment.
* DPI interface changed to transfer more than 64bit. svOpenArrayHandle is used. tsim library compilation now requires verilator includes
* Compute is changed to use TensorLoad style of load uop.
* VME changed to generate/queue/respond to multiple simultaneous load requests
* Added SyncQueue with tests - Implementation uses sync memory to implement larger queues.

Code contributions to this PR were made by the following individuals (in alphabetical order): @suvadeep89, @stevenmburns, @pasqoc, @adavare, @sjain12intel, @aasorokiin, and @zhenkuny.

### CODE REVIEW

**Comment by aasorokiin** [comment]
Adding @tmoreau89 . I need help on deciding how to proceed further. Checks flow fails because VTA DPI modification require changes to TVM. A Verilator include directory is required in order to build new TSIM DPI communication library by TVM make flow. Those changes are needed to transfer more than 64bit data as sv arrays. There is a pull request in TVM "VTA cmake change to include Verilator header for building tsim library" #8797. It fails checks as no Verilator found by cmake.

**Comment by tmoreau89** [comment]
Thank you @aasorokiin  - CC-ing @vegaluisjose

File: hardware/chisel/src/main/scala/util/SyncQueue.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/util/SyncQueue.scala]: Can we name this queue just `SynqQueue`?

File: src/dpi/module.cc
Comment (vegaluisjose) [inline src/dpi/module.cc]: is the clang formatter catching this?

File: src/dpi/module.cc
Comment (vegaluisjose) [inline src/dpi/module.cc]: remove comments

File: hardware/chisel/src/main/scala/core/TensorUtil.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorUtil.scala]: remove comment

File: hardware/chisel/src/main/scala/core/TensorUtil.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorUtil.scala]: remove comment

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala]: remove comment

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala:634]: can we add an extra space to comments?

File: hardware/chisel/src/main/scala/core/TensorLoadSimple.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/TensorLoadSimple.scala]: remove comment

File: hardware/chisel/src/main/scala/core/LoadUopSimple.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/LoadUopSimple.scala]: remove comment

File: hardware/chisel/src/main/scala/core/FetchWideVME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/FetchWideVME.scala]: remove comment

File: hardware/chisel/src/main/scala/core/FetchWideVME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/FetchWideVME.scala]: remove comment

File: hardware/chisel/src/main/scala/core/Compute.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/core/Compute.scala]: remove comment

File: hardware/chisel/src/main/resources/verilog/VTAMemDPI.v
Comment (vegaluisjose) [inline hardware/chisel/src/main/resources/verilog/VTAMemDPI.v:41]: can we align the port names to this one, to maintain the previous style?

**Review by vegaluisjose — CHANGES_REQUESTED** [review]

File: hardware/chisel/src/main/scala/util/SyncQueue.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/util/SyncQueue.scala]: Yes, it doesn't matter much. Updated PR.

File: src/dpi/module.cc
Comment (aasorokiin) [reply src/dpi/module.cc]: If it is not run by TVM make, than it is not. Fixed long lines in a new PR

File: src/dpi/module.cc
Comment (aasorokiin) [reply src/dpi/module.cc]: Updated PR.

File: hardware/chisel/src/main/scala/core/TensorUtil.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorUtil.scala]: Updated PR.

File: hardware/chisel/src/main/scala/core/TensorUtil.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorUtil.scala]: Updated PR.

File: hardware/chisel/src/main/scala/core/LoadUopSimple.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/LoadUopSimple.scala]: Updated PR.

File: hardware/chisel/src/main/scala/core/FetchWideVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/FetchWideVME.scala]: Updated PR.

File: hardware/chisel/src/main/scala/core/FetchWideVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/FetchWideVME.scala]: Updated PR.

File: hardware/chisel/src/main/scala/core/Compute.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/Compute.scala]: Updated PR.

File: hardware/chisel/src/main/resources/verilog/VTAMemDPI.v
Comment (aasorokiin) [reply hardware/chisel/src/main/resources/verilog/VTAMemDPI.v:41]: Updated PR.

**Comment by aasorokiin** [comment]
Updated this request to comply to Chisel changes from  #33
Cannot quite get why the latest main with #33 doesn't compile lib for me (make in chisel directory). It complains: --top-name cannot be used. Proposed Makefile changes are in this PR. @ekiwi

**Comment by vegaluisjose** [comment]
I think I found the issue of the new chisel version, I just submitted #34 that should fix it.

**Comment by tmoreau89** [comment]
#34 was just merged, @aasorokiin please rebase

**Comment by aasorokiin** [comment]
Changes were addressed.  TSIM ci also should pass now. @vegaluisjose

**Comment by vegaluisjose** [comment]
could you push en empty commit to see if we can re-trigger CI?

Thanks! @aasorokiin

**Comment by aasorokiin** [comment]
I tried that, but checks keep failing. 
I cannot reproduce the problem. For me it looks like CI caching error. @vegaluisjose

**Comment by vegaluisjose** [comment]
I think I found what is causing the issue. Can you try to update the docker image numbers [here](https://github.com/apache/tvm-vta/blob/main/Jenkinsfile#L24-L26) with the following numbers [here](https://github.com/apache/tvm/blob/main/Jenkinsfile#L47-L53)

the change should be
```
ci_lint = "tvmai/ci-lint:v0.67"
ci_cpu = "tvmai/ci-cpu:v0.77"
ci_i386 = "tvmai/ci-i386:v0.73"
```

File: Jenkinsfile
Comment (vegaluisjose) [inline Jenkinsfile]: just keep the lint, cpu, and i386. No need to add the others.

**Review by vegaluisjose — CHANGES_REQUESTED** [review]

**Comment by vegaluisjose** [comment]
Oh I see, what is going on now. Can you do this docker update on a separate PR? we will merge that before this one, so it can take the changes.

**Comment by vegaluisjose** [comment]
Hey @aasorokiin , now that we have #36 merged...could you please rebase to see how this one goes?

**Comment by aasorokiin** [comment]
Is there any reason tsim test run got skipped? @vegaluisjose

**Comment by vegaluisjose** [comment]
Oh yeah, could you please remove the following?

https://github.com/apache/tvm-vta/blob/main/tests/scripts/docker_bash.sh#L70-L74

I believe we fixed this already

File: hardware/chisel/src/main/scala/interface/axi/AXI.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/interface/axi/AXI.scala]: btw, what is up with `ar.bits.id` and `w.bits.strb` here? by default the strobe mask is all ones. Also, `ar.bits.id` is not being assigned now.

File: hardware/chisel/src/main/scala/shell/VME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/shell/VME.scala]: remove line

File: hardware/chisel/src/main/scala/shell/VME.scala
Comment (vegaluisjose) [inline hardware/chisel/src/main/scala/shell/VME.scala]: remove line

**Review by vegaluisjose — CHANGES_REQUESTED** [review]

File: hardware/chisel/src/main/scala/interface/axi/AXI.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/interface/axi/AXI.scala]: id/strb should be handled by VME now. VME can send multiple read requests with different IDs. Strb is used by TensorStore write to mark data of valid tensors in case when AXI data is wider than tensor data.

Comment (vegaluisjose) [reply hardware/chisel/src/main/scala/interface/axi/AXI.scala]: Sure but why changing the defaults? Isn't better to let these modules to customize (override) these values?

Also, if it is done for reads `ar` why it is not done for writes `aw`?

Comment (aasorokiin) [reply hardware/chisel/src/main/scala/interface/axi/AXI.scala]: IMHO, those are not defaults. They were set in setConst. That expected no change of value.
VME write path did not get changed other than supporting wider data.  It was not identified critical for hiding memory access latency. DE10 config AXI burst of 16 pulses was short enough to make hiding burst-to-burst read latency important. Now even 1 pulse bursts can potentially transfer read data without gaps.

File: hardware/chisel/src/main/scala/shell/VME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/shell/VME.scala]: fixed in PR

File: hardware/chisel/src/main/scala/shell/VME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/shell/VME.scala]: fixed in PR

File: hardware/chisel/src/main/scala/interface/axi/AXI.scala
Comment (vegaluisjose) [reply hardware/chisel/src/main/scala/interface/axi/AXI.scala]: Sure, the original intent of this method, back when we had support for one small board/configuration, was to use this as a mechanism to derive other params in the protocol. However, things have obviously evolved quite a bit since then.

What if we do the same for both reads/writes in terms of `id` so future contributors know that they have to handle `id` and `strb`?

Maybe we can even add comments on top of that `setConst` method saying what they should implement and we can point to that documentation later if needed.

For the record, these suggestions are intended to help future contributions. We really appreciate all your efforts on this!

Comment (aasorokiin) [reply hardware/chisel/src/main/scala/interface/axi/AXI.scala]: I moved write id to VME. Added comments. Check PR

**Review by vegaluisjose — APPROVED** [review]
Alright this looks like ready to go on my end, thanks a lot @aasorokiin !

do you have something to add? @tmoreau89

File: hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala
Comment (tmoreau89) [inline hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala]: please remove comment

File: hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala
Comment (tmoreau89) [inline hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala]: please remove comment

File: hardware/dpi/tsim_device.cc
Comment (tmoreau89) [inline hardware/dpi/tsim_device.cc]: tabs/spacing is off

File: hardware/dpi/tsim_device.cc
Comment (tmoreau89) [inline hardware/dpi/tsim_device.cc]: tabs/spacing is off

File: hardware/chisel/src/main/scala/core/LoadUop.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/LoadUop.scala]: clean up comments

File: hardware/chisel/src/main/scala/core/TensorLoadSimple.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadSimple.scala:240]: remove comment line above l240

File: hardware/chisel/src/test/scala/unittest/SyncQueue2PortMemTest.scala
Comment (tmoreau89) [inline hardware/chisel/src/test/scala/unittest/SyncQueue2PortMemTest.scala]: cleanup

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: typo

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala:86]: typo

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: typo: xas

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: enumerate

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala:729]: add new line

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (tmoreau89) [inline hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala]: remove comment above

**Comment by tmoreau89** [comment]
Thanks @vegaluisjose for reviewing the PR and ensuring that the CI tests go green, much appreciated. While going through the PR there were a few things that I'd like to address before we merge:

- Still quite a few comments that need to be cleaned up
- The comments could use a good spell check pass, many typos in some files - perhaps it could be automated via the preferred IDE
- Overall I think this is not blocking the merge, but we could extend the scala linter to address consistent comment formatting etc.

File: hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala
Comment (aasorokiin) [reply hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala]: fixed in PR

File: hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala
Comment (aasorokiin) [reply hardware/chisel/src/test/scala/unittest/SyncQueueTest.scala]: fixed in PR

File: hardware/dpi/tsim_device.cc
Comment (aasorokiin) [reply hardware/dpi/tsim_device.cc]: fixed in PR

File: hardware/dpi/tsim_device.cc
Comment (aasorokiin) [reply hardware/dpi/tsim_device.cc]: fixed in PR

File: hardware/chisel/src/main/scala/core/LoadUop.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/LoadUop.scala]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadSimple.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadSimple.scala:240]: fixed in PR

File: hardware/chisel/src/test/scala/unittest/SyncQueue2PortMemTest.scala
Comment (aasorokiin) [reply hardware/chisel/src/test/scala/unittest/SyncQueue2PortMemTest.scala]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala:86]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadNarrowVME.scala]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala:729]: fixed in PR

File: hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala
Comment (aasorokiin) [reply hardware/chisel/src/main/scala/core/TensorLoadWideVME.scala]: fixed in PR

**Comment by aasorokiin** [comment]
Is there anything else expected form me? @tmoreau89 @vegaluisjose

**Comment by tmoreau89** [comment]
Hi @aasorokiin  - thanks for the ping and addressing the comments.

**Comment by tmoreau89** [comment]
Thanks @vegaluisjose @aasorokiin and Intel team - the PR has been merged.

**Pull request merged**
